File: lib/sidereon/ccsds/tdm.ex
defmodule Sidereon.CCSDS.TDM do
  @moduledoc """
  Parse and encode CCSDS Tracking Data Messages (TDM) in KVN format.

  Date/time fields are preserved as raw strings. Data record values carry both
  the parsed `float` and the original decimal token so KVN round-trips do not
  rewrite measurement text.
  """

  alias Sidereon.NIF

  defmodule Field do
    @moduledoc """
    A preserved KVN key/value field.
    """

    @enforce_keys [:key, :value]
    defstruct [:key, :value]

    @type t :: %__MODULE__{key: String.t(), value: String.t()}
  end

  defmodule Observable do
    @moduledoc """
    Parsed TDM observable family.
    """

    @enforce_keys [:kind]
    defstruct [:kind, :participant, :name]

    @type t :: %__MODULE__{
            kind:
              :range
              | :doppler_instantaneous
              | :doppler_integrated
              | :receive_freq
              | :transmit_freq
              | :transmit_freq_rate
              | :angle_1
              | :angle_2
              | :other,
            participant: non_neg_integer() | nil,
            name: String.t() | nil
          }
  end

  defmodule Scalar do
    @moduledoc """
    A numeric TDM data value and its exact source token.
    """

    @enforce_keys [:text, :value]
    defstruct [:text, :value]

    @type t :: %__MODULE__{text: String.t(), value: float()}
  end

  defmodule DataRecord do
    @moduledoc """
    One time-tagged TDM tracking data record.
    """

    @enforce_keys [:observable, :keyword, :epoch, :value, :unit]
    defstruct [:observable, :keyword, :epoch, :value, :unit]

    @type t :: %__MODULE__{
            observable: Observable.t(),
            keyword: String.t(),
            epoch: String.t(),
            value: Scalar.t(),
            unit: String.t()
          }
  end

  defmodule DataSection do
    @moduledoc """
    TDM data block containing comments and records.
    """

    defstruct comments: [], records: []

    @type t :: %__MODULE__{comments: [String.t()], records: [DataRecord.t()]}
  end

  defmodule Participant do
    @moduledoc """
    One named TDM tracking participant.
    """

    @enforce_keys [:index, :name]
    defstruct [:index, :name]

    @type t :: %__MODULE__{index: non_neg_integer(), name: String.t()}
  end

  defmodule Path do
    @moduledoc """
    Parsed TDM signal path.
    """

    @enforce_keys [:key, :participants]
    defstruct [:key, :index, :participants]

    @type t :: %__MODULE__{
            key: String.t(),
            index: non_neg_integer() | nil,
            participants: [non_neg_integer()]
          }
  end

  defmodule Metadata do
    @moduledoc """
    Metadata block for one TDM segment.
    """

    defstruct comments: [],
              fields: [],
              participants: [],
              mode: nil,
              paths: [],
              timetag_ref: nil,
              time_system: nil,
              range_units: "km"

    @type t :: %__MODULE__{
            comments: [String.t()],
            fields: [Field.t()],
            participants: [Participant.t()],
            mode: String.t() | nil,
            paths: [Path.t()],
            timetag_ref: String.t() | nil,
            time_system: String.t() | nil,
            range_units: String.t()
          }
  end

  defmodule Segment do
    @moduledoc """
    One TDM metadata/data segment.
    """

    @enforce_keys [:metadata, :data]
    defstruct [:metadata, :data]

    @type t :: %__MODULE__{metadata: Metadata.t(), data: DataSection.t()}
  end

  @typedoc "Failure reason from TDM parsing or encoding."
  @type error ::
          :missing_version
          | :no_segments
          | {:section, {String.t(), String.t()}}
          | {:malformed_line, {String.t(), String.t()}}
          | {:malformed_record, {String.t(), String.t()}}
          | {:invalid_field, {String.t(), String.t()}}

  defstruct version: "2.0",
            comments: [],
            creation_date: nil,
            originator: nil,
            message_id: nil,
            header_fields: [],
            segments: []

  @type t :: %__MODULE__{
          version: String.t(),
          comments: [String.t()],
          creation_date: String.t() | nil,
          originator: String.t() | nil,
          message_id: String.t() | nil,
          header_fields: [Field.t()],
          segments: [Segment.t()]
        }

  @doc """
  Parse a TDM KVN document.
  """
  @spec parse(String.t()) :: {:ok, t()} | {:error, error()}
  def parse(text) when is_binary(text), do: parse_kvn(text)

  @doc """
  Parse a TDM KVN document explicitly.
  """
  @spec parse_kvn(String.t()) :: {:ok, t()} | {:error, error()}
  def parse_kvn(text) when is_binary(text) do
    text
    |> NIF.tdm_parse_kvn()
    |> from_nif_result()
  end

  @doc """
  Encode a TDM as KVN text.
  """
  @spec encode(t()) :: {:ok, String.t()} | {:error, error()}
  def encode(%__MODULE__{} = tdm), do: encode_kvn(tdm)

  @doc """
  Encode a TDM as KVN text explicitly.
  """
  @spec encode_kvn(t()) :: {:ok, String.t()} | {:error, error()}
  def encode_kvn(%__MODULE__{} = tdm) do
    tdm
    |> to_fields()
    |> NIF.tdm_encode_kvn()
    |> encoded_result()
  end

  defp from_nif_result({:ok, fields}), do: {:ok, from_fields(fields)}
  defp from_nif_result({:error, reason, detail}), do: {:error, error(reason, detail)}

  defp encoded_result({:ok, text}), do: {:ok, text}
  defp encoded_result({:error, reason, detail}), do: {:error, error(reason, detail)}

  defp error(reason, nil), do: reason
  defp error(reason, detail), do: {reason, detail}

  defp from_fields(fields) do
    %__MODULE__{
      version: fields.version,
      comments: fields.comments,
      creation_date: fields.creation_date,
      originator: fields.originator,
      message_id: fields.message_id,
      header_fields: Enum.map(fields.header_fields, &field_from_fields/1),
      segments: Enum.map(fields.segments, &segment_from_fields/1)
    }
  end

  defp field_from_fields(fields), do: %Field{key: fields.key, value: fields.value}

  defp observable_from_fields(fields) do
    %Observable{
      kind: String.to_atom(fields.kind),
      participant: fields.participant,
      name: fields.name
    }
  end

  defp scalar_from_fields(fields), do: %Scalar{text: fields.text, value: fields.value}

  defp record_from_fields(fields) do
    %DataRecord{
      observable: observable_from_fields(fields.observable),
      keyword: fields.keyword,
      epoch: fields.epoch,
      value: scalar_from_fields(fields.value),
      unit: fields.unit
    }
  end

  defp data_section_from_fields(fields) do
    %DataSection{
      comments: fields.comments,
      records: Enum.map(fields.records, &record_from_fields/1)
    }
  end

  defp participant_from_fields(fields), do: %Participant{index: fields.index, name: fields.name}

  defp path_from_fields(fields) do
    %Path{key: fields.key, index: fields.index, participants: fields.participants}
  end

  defp metadata_from_fields(fields) do
    %Metadata{
      comments: fields.comments,
      fields: Enum.map(fields.fields, &field_from_fields/1),
      participants: Enum.map(fields.participants, &participant_from_fields/1),
      mode: fields.mode,
      paths: Enum.map(fields.paths, &path_from_fields/1),
      timetag_ref: fields.timetag_ref,
      time_system: fields.time_system,
      range_units: fields.range_units
    }
  end

  defp segment_from_fields(fields) do
    %Segment{
      metadata: metadata_from_fields(fields.metadata),
      data: data_section_from_fields(fields.data)
    }
  end

  defp to_fields(%__MODULE__{} = tdm) do
    %{
      version: tdm.version,
      comments: tdm.comments,
      creation_date: tdm.creation_date,
      originator: tdm.originator,
      message_id: tdm.message_id,
      header_fields: Enum.map(tdm.header_fields, &field_to_fields/1),
      segments: Enum.map(tdm.segments, &segment_to_fields/1)
    }
  end

  defp field_to_fields(%Field{} = field), do: %{key: field.key, value: field.value}

  defp observable_to_fields(%Observable{} = observable) do
    %{
      kind: Atom.to_string(observable.kind),
      participant: observable.participant,
      name: observable.name
    }
  end

  defp scalar_to_fields(%Scalar{} = scalar), do: %{text: scalar.text, value: scalar.value / 1.0}

  defp record_to_fields(%DataRecord{} = record) do
    %{
      observable: observable_to_fields(record.observable),
      keyword: record.keyword,
      epoch: record.epoch,
      value: scalar_to_fields(record.value),
      unit: record.unit
    }
  end

  defp data_section_to_fields(%DataSection{} = data) do
    %{
      comments: data.comments,
      records: Enum.map(data.records, &record_to_fields/1)
    }
  end

  defp participant_to_fields(%Participant{} = participant) do
    %{index: participant.index, name: participant.name}
  end

  defp path_to_fields(%Path{} = path) do
    %{key: path.key, index: path.index, participants: path.participants}
  end

  defp metadata_to_fields(%Metadata{} = metadata) do
    %{
      comments: metadata.comments,
      fields: Enum.map(metadata.fields, &field_to_fields/1),
      participants: Enum.map(metadata.participants, &participant_to_fields/1),
      mode: metadata.mode,
      paths: Enum.map(metadata.paths, &path_to_fields/1),
      timetag_ref: metadata.timetag_ref,
      time_system: metadata.time_system,
      range_units: metadata.range_units
    }
  end

  defp segment_to_fields(%Segment{} = segment) do
    %{
      metadata: metadata_to_fields(segment.metadata),
      data: data_section_to_fields(segment.data)
    }
  end
end

File: native/sidereon_nif/src/tdm.rs
//! Rustler boundary for CCSDS Tracking Data Message KVN.
//!
//! This file mirrors the canonical core TDM structs as `NifMap` terms. Parsing,
//! validation, unit assignment, and KVN serialization remain in
//! `sidereon_core::astro::tdm`.

use rustler::{Encoder, Env, Term};
use sidereon_core::astro::tdm::{
    self as core_tdm, Tdm, TdmDataRecord, TdmDataSection, TdmError, TdmField, TdmInputErrorKind,
    TdmMetadata, TdmObservable, TdmParticipant, TdmPath, TdmScalar, TdmSegment, TdmUnit,
};

mod atoms {
    rustler::atoms! {
        ok,
        error,
        missing_version,
        no_segments,
        section,
        malformed_line,
        malformed_record,
        invalid_field
    }
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmFieldTerm {
    key: String,
    value: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmObservableTerm {
    kind: String,
    participant: Option<u8>,
    name: Option<String>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmScalarTerm {
    text: String,
    value: f64,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmDataRecordTerm {
    observable: TdmObservableTerm,
    keyword: String,
    epoch: String,
    value: TdmScalarTerm,
    unit: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmDataSectionTerm {
    comments: Vec<String>,
    records: Vec<TdmDataRecordTerm>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmParticipantTerm {
    index: u8,
    name: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmPathTerm {
    key: String,
    index: Option<u8>,
    participants: Vec<u8>,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmMetadataTerm {
    comments: Vec<String>,
    fields: Vec<TdmFieldTerm>,
    participants: Vec<TdmParticipantTerm>,
    mode: Option<String>,
    paths: Vec<TdmPathTerm>,
    timetag_ref: Option<String>,
    time_system: Option<String>,
    range_units: String,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmSegmentTerm {
    metadata: TdmMetadataTerm,
    data: TdmDataSectionTerm,
}

#[derive(Debug, Clone, rustler::NifMap)]
struct TdmTerm {
    version: String,
    comments: Vec<String>,
    creation_date: Option<String>,
    originator: Option<String>,
    message_id: Option<String>,
    header_fields: Vec<TdmFieldTerm>,
    segments: Vec<TdmSegmentTerm>,
}

fn field_term(value: TdmField) -> TdmFieldTerm {
    TdmFieldTerm {
        key: value.key,
        value: value.value,
    }
}

fn field(value: TdmFieldTerm) -> TdmField {
    TdmField {
        key: value.key,
        value: value.value,
    }
}

fn observable_term(value: TdmObservable) -> TdmObservableTerm {
    match value {
        TdmObservable::Range => TdmObservableTerm {
            kind: "range".to_string(),
            participant: None,
            name: None,
        },
        TdmObservable::DopplerInstantaneous => TdmObservableTerm {
            kind: "doppler_instantaneous".to_string(),
            participant: None,
            name: None,
        },
        TdmObservable::DopplerIntegrated => TdmObservableTerm {
            kind: "doppler_integrated".to_string(),
            participant: None,
            name: None,
        },
        TdmObservable::ReceiveFreq { participant } => TdmObservableTerm {
            kind: "receive_freq".to_string(),
            participant,
            name: None,
        },
        TdmObservable::TransmitFreq { participant } => TdmObservableTerm {
            kind: "transmit_freq".to_string(),
            participant,
            name: None,
        },
        TdmObservable::TransmitFreqRate { participant } => TdmObservableTerm {
            kind: "transmit_freq_rate".to_string(),
            participant,
            name: None,
        },
        TdmObservable::Angle1 => TdmObservableTerm {
            kind: "angle_1".to_string(),
            participant: None,
            name: None,
        },
        TdmObservable::Angle2 => TdmObservableTerm {
            kind: "angle_2".to_string(),
            participant: None,
            name: None,
        },
        TdmObservable::Other(name) => TdmObservableTerm {
            kind: "other".to_string(),
            participant: None,
            name: Some(name),
        },
    }
}

fn observable(value: TdmObservableTerm) -> TdmObservable {
    match value.kind.as_str() {
        "range" => TdmObservable::Range,
        "doppler_instantaneous" => TdmObservable::DopplerInstantaneous,
        "doppler_integrated" => TdmObservable::DopplerIntegrated,
        "receive_freq" => TdmObservable::ReceiveFreq {
            participant: value.participant,
        },
        "transmit_freq" => TdmObservable::TransmitFreq {
            participant: value.participant,
        },
        "transmit_freq_rate" => TdmObservable::TransmitFreqRate {
            participant: value.participant,
        },
        "angle_1" => TdmObservable::Angle1,
        "angle_2" => TdmObservable::Angle2,
        "other" => TdmObservable::Other(value.name.unwrap_or_default()),
        other => TdmObservable::Other(other.to_string()),
    }
}

fn unit_label(value: TdmUnit) -> String {
    value.as_str().to_string()
}

fn unit(value: String) -> TdmUnit {
    match value.as_str() {
        "km" => TdmUnit::Kilometers,
        "s" => TdmUnit::Seconds,
        "RU" => TdmUnit::RangeUnits,
        "km/s" => TdmUnit::KilometersPerSecond,
        "Hz" => TdmUnit::Hertz,
        "Hz/s" => TdmUnit::HertzPerSecond,
        "deg" => TdmUnit::Degrees,
        "dBW" => TdmUnit::DecibelWatts,
        "dBHz" => TdmUnit::DecibelHertz,
        "m**2" => TdmUnit::SquareMeters,
        "m" => TdmUnit::Meters,
        "s/s" => TdmUnit::SecondsPerSecond,
        "%" => TdmUnit::Percent,
        "K" => TdmUnit::Kelvin,
        "hPa" => TdmUnit::Hectopascals,
        "TECU" => TdmUnit::TotalElectronContentUnits,
        "n/a" => TdmUnit::Dimensionless,
        other => TdmUnit::Unknown(other.to_string()),
    }
}

fn scalar_term(value: TdmScalar) -> TdmScalarTerm {
    TdmScalarTerm {
        text: value.text,
        value: value.value,
    }
}

fn scalar(value: TdmScalarTerm) -> TdmScalar {
    TdmScalar {
        text: value.text,
        value: value.value,
    }
}

fn record_term(value: TdmDataRecord) -> TdmDataRecordTerm {
    TdmDataRecordTerm {
        observable: observable_term(value.observable),
        keyword: value.keyword,
        epoch: value.epoch,
        value: scalar_term(value.value),
        unit: unit_label(value.unit),
    }
}

fn record(value: TdmDataRecordTerm) -> TdmDataRecord {
    TdmDataRecord {
        observable: observable(value.observable),
        keyword: value.keyword,
        epoch: value.epoch,
        value: scalar(value.value),
        unit: unit(value.unit),
    }
}

fn data_section_term(value: TdmDataSection) -> TdmDataSectionTerm {
    TdmDataSectionTerm {
        comments: value.comments,
        records: value.records.into_iter().map(record_term).collect(),
    }
}

fn data_section(value: TdmDataSectionTerm) -> TdmDataSection {
    TdmDataSection {
        comments: value.comments,
        records: value.records.into_iter().map(record).collect(),
    }
}

fn participant_term(value: TdmParticipant) -> TdmParticipantTerm {
    TdmParticipantTerm {
        index: value.index,
        name: value.name,
    }
}

fn participant(value: TdmParticipantTerm) -> TdmParticipant {
    TdmParticipant {
        index: value.index,
        name: value.name,
    }
}

fn path_term(value: TdmPath) -> TdmPathTerm {
    TdmPathTerm {
        key: value.key,
        index: value.index,
        participants: value.participants,
    }
}

fn path(value: TdmPathTerm) -> TdmPath {
    TdmPath {
        key: value.key,
        index: value.index,
        participants: value.participants,
    }
}

fn metadata_term(value: TdmMetadata) -> TdmMetadataTerm {
    TdmMetadataTerm {
        comments: value.comments,
        fields: value.fields.into_iter().map(field_term).collect(),
        participants: value
            .participants
            .into_iter()
            .map(participant_term)
            .collect(),
        mode: value.mode,
        paths: value.paths.into_iter().map(path_term).collect(),
        timetag_ref: value.timetag_ref,
        time_system: value.time_system,
        range_units: unit_label(value.range_units),
    }
}

fn metadata(value: TdmMetadataTerm) -> TdmMetadata {
    TdmMetadata {
        comments: value.comments,
        fields: value.fields.into_iter().map(field).collect(),
        participants: value.participants.into_iter().map(participant).collect(),
        mode: value.mode,
        paths: value.paths.into_iter().map(path).collect(),
        timetag_ref: value.timetag_ref,
        time_system: value.time_system,
        range_units: unit(value.range_units),
    }
}

fn segment_term(value: TdmSegment) -> TdmSegmentTerm {
    TdmSegmentTerm {
        metadata: metadata_term(value.metadata),
        data: data_section_term(value.data),
    }
}

fn segment(value: TdmSegmentTerm) -> TdmSegment {
    TdmSegment {
        metadata: metadata(value.metadata),
        data: data_section(value.data),
    }
}

fn tdm_term(value: Tdm) -> TdmTerm {
    TdmTerm {
        version: value.version,
        comments: value.comments,
        creation_date: value.creation_date,
        originator: value.originator,
        message_id: value.message_id,
        header_fields: value.header_fields.into_iter().map(field_term).collect(),
        segments: value.segments.into_iter().map(segment_term).collect(),
    }
}

fn tdm(value: TdmTerm) -> Tdm {
    Tdm {
        version: value.version,
        comments: value.comments,
        creation_date: value.creation_date,
        originator: value.originator,
        message_id: value.message_id,
        header_fields: value.header_fields.into_iter().map(field).collect(),
        segments: value.segments.into_iter().map(segment).collect(),
    }
}

fn error_atom(error: &TdmError) -> rustler::Atom {
    match error {
        TdmError::MissingVersion => atoms::missing_version(),
        TdmError::NoSegments => atoms::no_segments(),
        TdmError::Section { .. } => atoms::section(),
        TdmError::MalformedLine { .. } => atoms::malformed_line(),
        TdmError::MalformedRecord { .. } => atoms::malformed_record(),
        TdmError::InvalidField { .. } => atoms::invalid_field(),
    }
}

fn error_kind_label(kind: TdmInputErrorKind) -> &'static str {
    match kind {
        TdmInputErrorKind::Missing => "missing",
        TdmInputErrorKind::FloatParse => "float_parse",
        TdmInputErrorKind::NonFinite => "non_finite",
        TdmInputErrorKind::NotPositive => "not_positive",
        TdmInputErrorKind::OutOfRange => "out_of_range",
        TdmInputErrorKind::InvalidIndex => "invalid_index",
        TdmInputErrorKind::UnknownKeyword => "unknown_keyword",
        TdmInputErrorKind::UnexpectedUnit => "unexpected_unit",
        TdmInputErrorKind::NonInteger => "non_integer",
        TdmInputErrorKind::Negative => "negative",
        TdmInputErrorKind::NegativeZero => "negative_zero",
        TdmInputErrorKind::UnitMismatch => "unit_mismatch",
        TdmInputErrorKind::DecimalMismatch => "decimal_mismatch",
    }
}

fn error_detail(error: &TdmError) -> Option<(String, String)> {
    match error {
        TdmError::InvalidField { field, kind } => {
            Some((field.clone(), error_kind_label(*kind).to_string()))
        }
        TdmError::Section { line, detail } => Some((line.to_string(), detail.to_string())),
        TdmError::MalformedLine { line, text } => Some((line.to_string(), text.clone())),
        TdmError::MalformedRecord { line, keyword } => Some((line.to_string(), keyword.clone())),
        TdmError::MissingVersion | TdmError::NoSegments => None,
    }
}

fn parse_result<'a>(env: Env<'a>, result: Result<Tdm, TdmError>) -> Term<'a> {
    match result {
        Ok(parsed) => (atoms::ok(), tdm_term(parsed)).encode(env),
        Err(error) => {
            let detail = error_detail(&error);
            (atoms::error(), error_atom(&error), detail).encode(env)
        }
    }
}

/// Parse a CCSDS TDM in KVN encoding.
#[rustler::nif(schedule = "DirtyCpu")]
fn tdm_parse_kvn<'a>(env: Env<'a>, text: String) -> Term<'a> {
    parse_result(env, core_tdm::parse_kvn(&text))
}

/// Serialize normalized TDM fields as CCSDS TDM KVN text.
#[rustler::nif(schedule = "DirtyCpu")]
fn tdm_encode_kvn<'a>(env: Env<'a>, fields: TdmTerm) -> Term<'a> {
    match core_tdm::encode_kvn(&tdm(fields)) {
        Ok(text) => (atoms::ok(), text).encode(env),
        Err(error) => {
            let detail = error_detail(&error);
            (atoms::error(), error_atom(&error), detail).encode(env)
        }
    }
}
